Add all CA certs to Kobo build so download of files works

GeekGoneOld opened the pull request:
Used the Mozilla CA trust list to validate download site
Fixes issue #871


<!--

Thank you for your interest in contributing to XCSoar! Please read the
following information to make it easier for us to review your changes.

We appreciate if you make sure to:

  - Document the changes (in the NEWS.txt file, and the manual when relevant)
  - Enable maintainer edits[1] (in case we need to help with the PR)
  - Check your commits and their messages (so they're all tidy and coherent)

[1] https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

-->


Brief summary of the changes
----------------------------

<!--
Please note that the commit messages is where detailed descriptions of the
changes should be made - this section is just for a brief summary/overview.
-->


Related issues and discussions
------------------------------

<!--
Please link any relevant issues or forum posts here, for reference.

If this PR resolves an existing issue, please write "Closes #1234" so that
the issue is closed automatically when this PR is merged.
-->

GeekGoneOld commented:
I have just changed the strategy a tiny bit.  
1. I used the download system to download and cache it.  Note I did this from libs.py right before curl because the file is versioned and if I did it in kobo.mk I would fear that it would never be updated until we had an outdated CA.  This way it is with a huge list of versioned files that are checked often.
2. I changed kobo.mk to use install to copy it into the correct folder
3. I can verify that it did not overwrite an existing file.  There are four other folders that it can be placed in:
        /etc/pki/tls/certs/ca-bundle.crt
        /usr/share/ssl/certs/ca-bundle.crt
        /usr/local/share/certs/ca-root-nss.crt
        /etc/ssl/cert.pem)
  If you have a preference for me to use one of these, let me know.
5. I could have altered the curl build to use any folder and name I want but I'd be afraid that could cause another target (e.g. android or windows) to fail.

Let me know if you would like me to change anything.

MaxKellermann commented on build/kobo.mk:
```diff
@@ -81,6 +81,7 @@ THIRDPARTY_TOOL_NAMES = simple_usbmodeswitch
 THIRDPARTY_TOOL_FILES = $(addprefix $(THIRDPARTY_LIBS_ROOT)/bin/,$(THIRDPARTY_TOOL_NAMES))
 
 KOBO_KERNEL_DIR = /opt/kobo/kernel
+CA_FILE != ls -v $(DOWNLOAD_DIR)/cacert*.pem | tail -n  1
```
This looks... fragile.
For example, it breaks when the file does not already exist.

GeekGoneOld replied:
removed

MaxKellermann commented on build/kobo.mk:
```diff
@@ -104,6 +105,7 @@ $(TARGET_OUTPUT_DIR)/KoboRoot.tgz: $(XCSOAR_BIN) \
 	$(Q)install -m 0644 $(topdir)/kobo/inittab $(@D)/KoboRoot/etc
 	$(Q)install -m 0644 $(topdir)/kobo/udev.rules $(@D)/KoboRoot/etc/udev/rules.d/99-xcsoar.rules
 	$(Q)install -m 0644 $(BITSTREAM_VERA_FILES) $(@D)/KoboRoot/opt/xcsoar/share/fonts
+	$(Q)install -T -m 0644 $(CA_FILE) $(@D)/KoboRoot/etc/ssl/certs/ca-certificates.crt
```
This will split obscure error messages when there's no CA file in the download directory.

GeekGoneOld replied:
> This looks... fragile.

removed.

> This will split obscure error messages when there's no CA file in the download directory.

Changed .tgz file to be dependent upon file in d/l directory.

> Why don't you use the same thing as the Boost download does:

Made like Boost.  Note this will require kobo.mk to be updated from time to time to get the latest CA version (as with boost and many other files).

GeekGoneOld replied:
Changed .tgz file to be dependent upon file in d/l directory.

MaxKellermann commented on build/python/build/libs.py:
```diff
@@ -176,6 +177,13 @@
     patches=abspath('lib/c-ares/patches'),
 )
 
+ca = download_and_verify(
```
This unconditionally downloads the CA file, even if you're compiling a target that doesn't use them.
Why don't you use the same thing as the Boost download does:
https://github.com/XCSoar/XCSoar/blob/812d86055bcaf69d830fa005116c5de8815f4a33/build/libboost.mk#L12-L14

GeekGoneOld replied:
> This looks... fragile.

removed.

> This will split obscure error messages when there's no CA file in the download directory.

Changed .tgz file to be dependent upon file in d/l directory.

> Why don't you use the same thing as the Boost download does:

Made like Boost.  Note this will require kobo.mk to be updated from time to time to get the latest CA version (as with boost and many other files).

GeekGoneOld replied:
Just nudging to see if I've satisfied all your comments.

GeekGoneOld commented:
> This looks... fragile.

removed.

> This will split obscure error messages when there's no CA file in the download directory.

Changed .tgz file to be dependent upon file in d/l directory.

> Why don't you use the same thing as the Boost download does:

Made like Boost.  Note this will require kobo.mk to be updated from time to time to get the latest CA version (as with boost and many other files).

GeekGoneOld commented:
@folken or @MaxKellermann 
I don't understand the failure of the UNIX build.  All I did was rebase because my commit for the PR was 3 commits (all by Max) behind.  Don't know why it failed. Details above don't mean all that much to me. My PR is Kobo specific so I wonder whether something went wrong with the Unix build 9 days ago (the base for my PR).  Let me know if there is anything I should look at.

GeekGoneOld commented:
Yup, all checks passed this time so the previous failure must have been a transient failure during the build check.

Hopefully my PR meets the teams requirements but if not, let me know what you wish altered.

MaxKellermann commented:
No time right now for review, sorry. Will do next week.

GeekGoneOld commented:
@MaxKellermann @lordfolken Are you waiting for me on this PR?  It says "changes requested" and maybe that is what you are waiting for but I have made changes to address all concerns and don't have a clue how to address the "changes requested" flag.  I'm hoping you can approve and pull this or tell me what changes I must address.

lordfolken commented:
I updated it to the latest certificate file. Also changed the checksum to an md5. (Why are we still using md5?)
And rebased to current master.

I don't have a kobo to test on, so I would be very happy if someone could  test this. 
The KoboRoot.tgz.zip is downloadable on the github action run. 

@cmoetzing could you test this please?

GeekGoneOld commented:
> @cmoetzing could you test this please?

You probably meant @GeekGoneOld as I authored the issue and PR that you updated.  Unfortunately, I'm unable to test for 3 weeks so if @cmoetzing can test it I would be thrilled.  I had tried it prior to your changes and it worked perfectly.

> The KoboRoot.tgz.zip is downloadable on the github action run.

I MIGHT be able to test it if you explain how I get to the github action run and download the .zip.  I've submitted PRs but never heard of action run.

cmoetzing commented:
@lordfolken @GeekGoneOld I installed the version from the action run and tried to download the UK airspaces. Unfortunately I got an error.

`[2022-05-18T17:53:17Z] Failed to create x!h4oard/XCSoarData/UK_Airspace.txt: No such file or directory`

Should I try another file?

lordfolken commented:
@cmoetzing thank you for testing. 
```ini
name=BRA_SOUTH.xcm
uri=https://download.xcsoar.org/source/map/region/BRA_SOUTH.xcm
type=map
area=br
update=2021-12-28
```
Is hosted on https so that is already a good sign. 
Do you think the errors you had where connectivity issues?

@GeekGoneOld I wanted that someone else does the testing, so we have a good overview whether we haven't forgotten to add something from the developers environment.
The github action run, you can f1ind by going to the "actions" tab on the repository, select the kobo-run for this pr, and there you will find "artefacts" https://github.com/XCSoar/XCSoar/actions/runs/234490515 (you need to be signed into github)

cmoetzing commented:
> @cmoetzing thank you for testing.
> 
> ```ini
> name=BRA_SOUTH.xcm
> uri=https://download.xcsoar.org/source/map/region/BRA_SOUTH.xcm
> type=map
> area=br
> update=2021-12-28
> ```
> 
> Is hosted on https so that is already a good sign. Do you think the errors you had where connectivity issues?
> 
> @GeekGoneOld I wanted that someone else does the testing, so we have a good overview whether we haven't forgotten to add something from the developers environment. The github action run, you can f1ind by going to the "actions" tab on the repository, select the kobo-run for this pr, and there you will find "artefacts" https://github.com/XCSoar/XCSoar/actions/runs/234490515 (you need to be signed into github)

I think I might have been too impatient clicking "Download" on the entries when it was showing "In Queue".

lordfolken commented:
@GeekGoneOld want to add yourself to AUTHORS.txt?

GeekGoneOld commented:
I assume this will allow people to contact me better if needed. Sounds fair. 

Keith Laidlaw
***@***.***

> On May 19, 2022, at 7:22 PM, Philipp Wollschlegel ***@***.***> wrote:
> 
> ﻿
> @GeekGoneOld want to add yourself to AUTHORS.txt?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.
